pkg/kubelet/util/swap/swap_util.go

```diff
@@ -0,0 +1,125 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package swap
+
+import (
+	"bytes"
+	"fmt"
+	"k8s.io/klog/v2"
+	"k8s.io/mount-utils"
+	"os"
+	sysruntime "runtime"
+	"strings"
+	"sync"
+)
+
+var (
+	tmpfsNoswapOptionSupported        bool
+	tmpfsNoswapOptionAvailabilityOnce sync.Once
+	swapOn                            bool
+	swapOnErr                         error
+	swapOnOnce                        sync.Once
+	rootDir                           string
+)
+
+const TmpfsNoswapOption = "noswap"
+
+// InitializeRootDir gets a path for a base directory, to which we have write permission.
+// This path will be later use for writing temporary files / mount points / etc.
+func InitializeRootDir(path string) {
+	rootDir = path
+}
+
+func IsTmpfsNoswapOptionSupported(mounter mount.Interface) bool {
+	isTmpfsNoswapOptionSupportedHelper := func() bool {
+		if sysruntime.GOOS == "windows" {
+			return false
+		}
+
+		baseMountDir := rootDir
+		if baseMountDir == "" {
+			klog.ErrorS(fmt.Errorf("swap utils root dir is not initialized, falling back to /tmp"), "")
+		}
+
+		mountDir, err := os.MkdirTemp(baseMountDir, "tmpfs-noswap-test-")
+		if err != nil {
+			klog.InfoS("error creating dir to test if tmpfs noswap is enabled. Assuming not supported", "mount path", mountDir, "error", err)
+			return false
+		}
+
+		defer func() {
+			err = os.RemoveAll(mountDir)
+			if err != nil {
+				klog.ErrorS(err, "error removing test tmpfs dir", "mount path", mountDir)
+			}
+		}()
+
+		err = mounter.MountSensitiveWithoutSystemd("tmpfs", mountDir, "tmpfs", []string{TmpfsNoswapOption}, nil)
+		if err != nil {
+			klog.InfoS("error mounting tmpfs with the noswap option. Assuming not supported", "error", err)
+			return false
+		}
+
+		err = mounter.Unmount(mountDir)
+		if err != nil {
+			klog.ErrorS(err, "error unmounting test tmpfs dir", "mount path", mountDir)
+		}
+
+		return true
+	}
+
+	tmpfsNoswapOptionAvailabilityOnce.Do(func() {
+		tmpfsNoswapOptionSupported = isTmpfsNoswapOptionSupportedHelper()
+	})
+
+	return tmpfsNoswapOptionSupported
+}
+
+// IsSwapOn detects whether swap in enabled on the system by inspecting
+// /proc/swaps. If the file does not exist, an os.NotFound error will be returned.
+// If running on windows, swap is assumed to always be false.
```

Comment:
Could we add the expected content of /proc/swaps here for readability, or better yet, with a unit test?